<br> 

## <Center>  Improving Algorithmic Performance</Center>  
## <Center>  Performance Testing and Algorithm Optimization  </Center>  

## <Center>  Rodrigo Medina </Center>  

  <br>
  <h1><left> Index  </Center>  </left>
  <br>


 
###     1. INTRODUCTION
###     2. PROBLEM DEFINITION
###     3. VARIABLE DEFINITIONS AND INTITIAL CONFIGURATIONS
###     4. SOLUTION 1
###     5. SOLUTION 2
###     6. PERFORMANCE COMPARATION  
#

 

<h2><center> 1. Introduction</center></h2>


<left><b> The primary goal of this project is to illustrate the solution of a given problem through the use of two distinct algorithms, emphasizing the difference in performance between these two different solutions.This will highlight the significance of performance testing and algorithm tuning, emphasizing the necessity to build solutions that not only effectively solve problems but also prioritize efficiency in terms of time and resource utilization. </b></left>
   <br> <br>
   
   

<h2><center>Results Highlights </center></h2>


<left><b> The project will demonstrate how speed in computation time is reduced by 1500x to 2500x, when employing solution 2 compared to solution 1. This scenario highlights how an alternative approach, involving intricate mathematical reasoning can present a far more efficient solution to the approach of evaluating all possible scenarios.    </b></left>
   <br> <br>
   
  

<br> <br>    
<h2><center> 2. Problem Definition</center></h2>



  <p><left> Design a function that takes an array of integers, Array A, as input and generates two subarrays, Sub Array B and Sub Array C, such that the sum of Sub Array B and SubArray C´s elements,  exactly matches the elements in Array A. The function should return the Sub Array B with the shortest possible length and with the highest possible sum of its elements. The sum of elements of Sub Array B must be greater than the sum of elements of Sub Array C, and Sub Array B cannot have more elements than Sub Array C
     </left><br>
     
     


 <p><left> Example: 
     </left> 
      <p><left><b> Array A =</b>  [5,2,4,6,1,3,14] , <b> Sub Array B = ? , <b> Sub Array C = ?
     </left><br><br>    
    
               
  <p><left>For lenght 1 of Sub Array B no instance of Subarray B meets the required conditions, for lenght 2 there are 3 instances when the sum of Sub Array B is greater than the Sum of Sub Array C and Sub Array B + Sub Array C = Array A:
     </left><br>  
      
   <p><left>    
      <b> Sub Array B = </b> [14,4] <b>, Sum  Array B =</b> 18, <b>Sub Array C =</b>  [5,2,6,1,3] <b>Sum Array C = </b> 17     </left><br> 
       
   <p><left>    
       <b> Sub Array B =</b> [14,5] <b>, Sum  Array B =</b> 19 , <b>Sub Array C =</b>  [6,2,4,1,3] <b>Sum Array C =</b>  16     </left><br>  
              
   <p><left>    
       <b> Sub Array B =</b> [14,6] <b>, Sum  Array B =</b> 20 , <b>Sub Array C = </b> [5,2,4,1,3] <b>Sum Array C =</b>  15     </left><br> 
       <br>  
                      
  <p><left>Out of the 3 possible combinations [14,6],[14,5],[14,4]: <br> <br>  
      <b> Sub Array B = [14,6] is the combination with the highest sum of its elements (20) and shortest lenght (2 elements), this should be the answer of the function</b>
     </left><br>  

<br> <br>    
<h2><center> 3. Variable Definitions and Initial Conficurations</center></h2>



In [40]:
#Importing Required Packages
import time
import random

<p><left>    
      <b> Simple arrays to test in functions </b>      </left><br> 

In [41]:
simple_array_1 = [5,2,4,6,1,3,12]
simple_array_2 = [1,7,4,6,8,16,4,6]
simple_array_3 = [15,22,14,56,31]


<p><left>    
      <b> Complex arrays to test in functions:<br><br> Generating 3 arrays with 30,000 , 35,000 and 40,000 elements, populated with random numbers from 1 to 1000 </b>      </left><br> 

In [42]:
random_array = [random.randint(1, 1001) for _ in range(30000)]
arr = random_array

random_array2 = [random.randint(1, 1001) for _ in range(35000)]
arr2 = random_array2

random_array3 = [random.randint(1, 1001) for _ in range(50000)]
arr3 = random_array3

<br> <br>    
<h2><center> 4. Solution 1</center></h2>



 <p><left> Solution One calculates the half lenght of Array A, which is ((n-1)/2) for the case of odd lenght of Array A and (n)/2) for the case of even lenght of Array A.
     </left></p>
     
<p><left> Since Sub Array B should have less elements than Sub Array C , the function will evaluate all possible combinaitons of Sub Array B in lenghts from 1 to ((n-1)/2) or (n)/2), (Depending on odd or even lenght of Array A), if the lenght of Sub Array B was larger than ((n-1)/2) or (n)/2), then it would have more elements than Sub Array C and would not meet the intial conditions. All combinations of Sub Array A will be saved in array of arrays variable, then from that variable the array with the shortest lenght and highest sum will be choosen.
     </left></p>

<p><left>To ensure Sub Array A has the highest possible sum Array A will be sorted in Descending Order, and from that sorted form of Array A, Sub Array B and Sub Array C will be selected.
     </left></p>   

In [43]:
def min_lenght_2_arrays_version_1(arr):
    start_time = time.time()
    arr = sorted(arr, reverse=True)
    n = len(arr)

    if n % 2 == 1:
        loop_data = []
        for i in range((int((n-1)/2)),0,-1):
            arr_subset_a = arr[:(i)]
            arr_subset_b = arr[(i):]
            sum_a = sum(arr_subset_a)
            sum_b = sum(arr_subset_b)
            
            loop_data.append({
                'sum_a': sum_a,
                'sum_b': sum_b,
                'arr_subset_a': arr_subset_a,
                'arr_subset_b': arr_subset_b
            })

        higher_sum_a_data = [data for data in loop_data if data['sum_a'] > data['sum_b']]
        result_data = min(higher_sum_a_data, key=lambda x: len(x['arr_subset_a']))
        end_time = time.time()
        run_time = end_time - start_time
        if n > 10:
            return "Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:", len(result_data['arr_subset_a']) , f"Run time: {round(run_time, 6)} seconds" ,"The resulting array is saved in object arr_subset_a "
        
        if n <= 10:
            
            return "Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:", len(result_data['arr_subset_a']) , f"Run time: {round(run_time, 6)} seconds" ,"The resulting array is saved in object arr_subset_a ","Array A:", arr, "Sub Array B:" , result_data['arr_subset_a'], "Sub Array C:" , result_data['arr_subset_b']
        
        
    if n % 2 == 0:
        loop_data = []
        for i in range((int((n)/2)),0,-1):
            arr_subset_a = arr[:(i)]
            arr_subset_b = arr[(i):]
            sum_a = sum(arr_subset_a)
            sum_b = sum(arr_subset_b)
            
            loop_data.append({
                'sum_a': sum_a,
                'sum_b': sum_b,
                'arr_subset_a': arr_subset_a,
                'arr_subset_b': arr_subset_b
            })

        higher_sum_a_data = [data for data in loop_data if data['sum_a'] > data['sum_b']]
        result_data = min(higher_sum_a_data, key=lambda x: len(x['arr_subset_a']))
        end_time = time.time()
        run_time = end_time - start_time
        
        if n > 10:
            return "Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:", len(result_data['arr_subset_a']) , f"Run time: {round(run_time, 6)} seconds" ,"The resulting array is saved in object arr_subset_a "
        
        if n <= 10:
            
            return "Solution 1 - The minimum length array whose sum of elements is greater than the rest of the elements is:", len(result_data['arr_subset_a']) , f"Run time: {round(run_time, 6)} seconds" ,"The resulting array is saved in object arr_subset_a ","Array A:", arr, "Sub Array B:" , result_data['arr_subset_a'], "Sub Array C:" , result_data['arr_subset_b']
        


<p><left>    
      <b> Testing Solution 1 in Simple Cases </b>      </left><br> 

In [44]:
print("Example 1 - ", "Simple array with " , len(simple_array_1) ," elements")
print("")
print(min_lenght_2_arrays_version_1(simple_array_1))
print("")
print("Example 2 - ", "Simple array with " , len(simple_array_2) ," elements")
print("")
print(min_lenght_2_arrays_version_1(simple_array_2))
print("")
print("Example 3 - ", "Simple array with " , len(simple_array_3) ," elements")
print("")
print(min_lenght_2_arrays_version_1(simple_array_3))


Example 1 -  Simple array with  7  elements

('Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 2, 'Run time: 0.0 seconds', 'The resulting array is saved in object arr_subset_a ', 'Array A:', [12, 6, 5, 4, 3, 2, 1], 'Sub Array B:', [12, 6], 'Sub Array C:', [5, 4, 3, 2, 1])

Example 2 -  Simple array with  8  elements

('Solution 1 - The minimum length array whose sum of elements is greater than the rest of the elements is:', 3, 'Run time: 0.0 seconds', 'The resulting array is saved in object arr_subset_a ', 'Array A:', [16, 8, 7, 6, 6, 4, 4, 1], 'Sub Array B:', [16, 8, 7], 'Sub Array C:', [6, 6, 4, 4, 1])

Example 3 -  Simple array with  5  elements

('Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 2, 'Run time: 0.0 seconds', 'The resulting array is saved in object arr_subset_a ', 'Array A:', [56, 31, 22, 15, 14], 'Sub Array B:', [56, 31], 'Sub Array C:', [22, 15, 14])


<p><left><b> Testing Solution 1 in Complex Cases </b></left><br> 

In [45]:
print("Example 4 - ", "array with " , len(arr) ," elements")
print("")
print(min_lenght_2_arrays_version_1(arr))
print("")
print("Example 5 - ", "array with " , len(arr2) ," elements")
print("")
print(min_lenght_2_arrays_version_1(arr2))
print("Example 6 - ", "array with " , len(arr3) ," elements")
print("")
print(min_lenght_2_arrays_version_1(arr3))


Example 4 -  array with  30000  elements

('Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 8820, 'Run time: 13.078528 seconds', 'The resulting array is saved in object arr_subset_a ')

Example 5 -  array with  35000  elements

('Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 10281, 'Run time: 17.646063 seconds', 'The resulting array is saved in object arr_subset_a ')
Example 6 -  array with  50000  elements

('Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 14622, 'Run time: 42.146186 seconds', 'The resulting array is saved in object arr_subset_a ')


<h3><left><b> Solution 1 Conclusion </b></left></h3>
    
<p><left> Solution 1 works correctly, however we can notice that as the number of elements in an Array A increases computation time increases significantly. Solutions in the complex cases take more than 20 seconds, and continue to increase non linearly as the number of elements increase
     </left></p>
     



<br> <br>    
<h2><center> 5. Solution 2</center></h2>



  <p><left> If we sort Array A in descending order and we start adding elements to Sub Arrray B, we garantuee we are adding the highest value possible combinations for Sub Array B, we are also starting from the smallest possible lenght for Sub Array B, so if we continue this process until the point that Sub Array B is larger than Sub Array C we will find the shortest lenght possible array in which the sum of elements is greater than the sum of elements in Sub Array C.
     </left><br>
     
  <p><left> The moment the Sum of Elements of Sub Array B is greater than half of the sum of all the elements in Array A will imply that the rest of the elements (Sub Array C) will be smaller than Sub Array B, using this logic the solution will add elements into Sub Array B and stop when the total sum of the elements in Sub Array B is gretaer than half of the sum of all the elements in Array A 
     </left><br>    
     

In [46]:
def min_lenght_2_arrays_version_2(arr):
    start_time = time.time() 
    arr = sorted(arr, reverse=True)
    mid_point = (sum(arr))/2
    counter = 0
    target_value = 0
    for i in arr:    
        if target_value > mid_point:
            break
        else:
            counter += 1
            target_value += i   
    arr_subset_c = arr[0:counter]
    arr_subset_d = arr[(counter):]                                                                  
    end_time = time.time()
    run_time = end_time - start_time
    
    if len(arr[0:counter]) > 10:
    
        return "Solution 2 - The minimum length array whose sum of elements is greater than the rest of the elements is:", len(arr_subset_c) , f"Run time: {round(run_time, 6)} seconds", "The resulting array is saved in object arr_subset_c  "

    if len(arr[0:counter]) <= 10:

        return "Solution 2 -The minimum length array whose sum of elements is greater than the rest of the elements is:", len(arr_subset_c)  , f"Run time: {round(run_time, 6)} seconds" ,"The resulting array is saved in object arr_subset_c ","Array A:", arr, "Sub Array B:" , arr_subset_c , "Sub Array C:" , arr_subset_d 
        
    

<p><left>    
      <b> Testing Solution 2 in Simple Cases </b>      </left><br> 

In [47]:
print("Example 1 - ", "Simple array with " , len(simple_array_1) ," elements")
print("")
print(min_lenght_2_arrays_version_2(simple_array_1))
print("")
print("Example 2 - ", "Simple array with " , len(simple_array_2) ," elements")
print("")
print(min_lenght_2_arrays_version_2(simple_array_2))
print("")
print("Example 3 - ", "Simple array with " , len(simple_array_3) ," elements")
print("")
print(min_lenght_2_arrays_version_2(simple_array_3))

Example 1 -  Simple array with  7  elements

('Solution 2 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 2, 'Run time: 0.0 seconds', 'The resulting array is saved in object arr_subset_c ', 'Array A:', [12, 6, 5, 4, 3, 2, 1], 'Sub Array B:', [12, 6], 'Sub Array C:', [5, 4, 3, 2, 1])

Example 2 -  Simple array with  8  elements

('Solution 2 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 3, 'Run time: 0.0 seconds', 'The resulting array is saved in object arr_subset_c ', 'Array A:', [16, 8, 7, 6, 6, 4, 4, 1], 'Sub Array B:', [16, 8, 7], 'Sub Array C:', [6, 6, 4, 4, 1])

Example 3 -  Simple array with  5  elements

('Solution 2 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 2, 'Run time: 0.0 seconds', 'The resulting array is saved in object arr_subset_c ', 'Array A:', [56, 31, 22, 15, 14], 'Sub Array B:', [56, 31], 'Sub Array C:', [22, 15, 14])


<p><left>    
      <b> Testing Solution 2 in Complex Cases </b>      </left><br> 

In [48]:
print("Example 4 - ", "array with " , len(arr) ," elements")
print("")
print(min_lenght_2_arrays_version_2(arr))
print("")
print("Example 5 - ", "array with " , len(arr2) ," elements")
print("")
print(min_lenght_2_arrays_version_2(arr2))
print("")
print("Example 6 - ", "array with " , len(arr3) ," elements")
print("")
print(min_lenght_2_arrays_version_2(arr3))
print("")

Example 4 -  array with  30000  elements

('Solution 2 - The minimum length array whose sum of elements is greater than the rest of the elements is:', 8820, 'Run time: 0.0 seconds', 'The resulting array is saved in object arr_subset_c  ')

Example 5 -  array with  35000  elements

('Solution 2 - The minimum length array whose sum of elements is greater than the rest of the elements is:', 10281, 'Run time: 0.012619 seconds', 'The resulting array is saved in object arr_subset_c  ')

Example 6 -  array with  50000  elements

('Solution 2 - The minimum length array whose sum of elements is greater than the rest of the elements is:', 14622, 'Run time: 0.016644 seconds', 'The resulting array is saved in object arr_subset_c  ')



<h3><left><b> Solution 2 Conclusion </b></left></h3>
    
<p><left> Solution 2 works correctly, and we can notice that the computation times for even the complex cases range between .01 and .02 seconds a very significant improvement in performance when compared to solution 1
     </left></p>
     

<br> <br>    
<h2><center> 6. Performance Comparation</center></h2>



  <p><left> Solution 1 and Solution 2 will be executed consecutively on three complex cases, and their computation times will be compared to have an understanding of the difference in performance between the two solutions
     </left><br>
     
 
     

In [49]:
print("Example 4 - ", "array with " , len(arr) ," elements")
print("")
print(min_lenght_2_arrays_version_1(arr))
print("")
print(min_lenght_2_arrays_version_2(arr))
print("")
print("Example 5 - ", "array with " , len(arr2) ," elements")
print("")
print(min_lenght_2_arrays_version_1(arr2))
print("")
print(min_lenght_2_arrays_version_2(arr2))
print("")
print("Example 6 - ", "array with " , len(arr3) ," elements")
print("")
print(min_lenght_2_arrays_version_1(arr3))
print("")
print(min_lenght_2_arrays_version_2(arr3))
print("")

Example 4 -  array with  30000  elements

('Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 8820, 'Run time: 13.155358 seconds', 'The resulting array is saved in object arr_subset_a ')

('Solution 2 - The minimum length array whose sum of elements is greater than the rest of the elements is:', 8820, 'Run time: 0.005229 seconds', 'The resulting array is saved in object arr_subset_c  ')

Example 5 -  array with  35000  elements

('Solution 1 -The minimum length array whose sum of elements is greater than the rest of the elements is:', 10281, 'Run time: 17.115891 seconds', 'The resulting array is saved in object arr_subset_a ')

('Solution 2 - The minimum length array whose sum of elements is greater than the rest of the elements is:', 10281, 'Run time: 0.010852 seconds', 'The resulting array is saved in object arr_subset_c  ')

Example 6 -  array with  50000  elements

('Solution 1 -The minimum length array whose sum of elements i

<h2><left><b> Example 4 Array with 30,000 Elements </b></left></h2>
    
<p><left> Solution 1 took 13.155358 seconds to obtain the result, Solution 2 took 0.005229 seconds to obtain the result.
     </left></p>
     

In [54]:
13.155358/0.005229

2515.845859628992

<h3><left><b> Solution 2 was 2515.84 times faster than solution 1 </b></left></h3>
    


<br><br><h2><left><b> Example 5 Array with 35,000 Elements </b></left></h2>
    
<p><left> Solution 1 took 17.115891 seconds to obtain the result, Solution 2 took  0.01085 seconds to obtain the result.
     </left></p>
     

In [56]:
17.115891/0.010852

1577.2107445632141

<h3><left><b> Solution 2 was 1577.21 times faster than solution 1 </b></left></h3>
    


<br> <br>  <h2><left><b> Example 6 Array with 50,000 Elements </b></left></h2>
    
<p><left> Solution 1 took 42.621385 seconds to obtain the result, Solution 2 took 0.024349 seconds to obtain the result.
     </left></p>
     

In [57]:
42.621385/0.024349

1750.4367735841306

<h3><left><b> Solution 2 was 1750.43 times faster than solution 1 </b></left></h3>
    
